verifier/verifier.go

```diff
@@ -0,0 +1,265 @@
+	"github.com/notaryproject/notation-go/internal/envelope"
+	"github.com/notaryproject/notation-go/notation"
+	"github.com/notaryproject/notation-go/plugin"
+	"github.com/notaryproject/notation-go/plugin/manager"
+	"github.com/notaryproject/notation-go/verifier/trustpolicy"
+)
+
+// verifier implements notation.Verifier
+type verifier struct {
+	TrustPolicyDoc *trustpolicy.Document
+	PluginManager  pluginManager
+}
+
+// pluginManager is for mocking in unit tests
+type pluginManager interface {
+	Get(ctx context.Context, name string) (*manager.Plugin, error)
+	Runner(name string) (plugin.Runner, error)
+}
+
+// New a verifier based on local file system
+func New() (notation.Verifier, error) {
+	// load trust policy
+	policyDocument, err := loadPolicyDocument()
+	if err != nil {
+		return nil, err
+	}
+	if err = policyDocument.Validate(); err != nil {
+		return nil, err
+	}
+
+	return &verifier{
+		TrustPolicyDoc: policyDocument,
+		PluginManager:  manager.New(dir.PluginFS()),
+	}, nil
+}
+
+// NewVerifier creates a new verifier given trustPolicy and pluginManager
+func NewVerifier(trustPolicy *trustpolicy.Document, pluginManager *manager.Manager) (notation.Verifier, error) {
+	if err := trustPolicy.Validate(); err != nil {
+		return nil, err
+	}
+	return &verifier{
+		TrustPolicyDoc: trustPolicy,
+		PluginManager:  pluginManager,
+	}, nil
+}
+
+// TrustPolicyDocument gets the validated trust policy document.
+func (v *verifier) TrustPolicyDocument() (*trustpolicy.Document, error) {
+	if err := v.TrustPolicyDoc.Validate(); err != nil {
+		return nil, err
+	}
+	return v.TrustPolicyDoc, nil
+}
+
+// Verify verifies the signature blob and returns the verified descriptor
+// upon successful verification.
+func (v *verifier) Verify(ctx context.Context, signature []byte, opts notation.VerifyOptions) (notation.Descriptor, *notation.VerificationOutcome, error) {
+	artifactRef := opts.ArtifactReference
+	envelopeMediaType := opts.SignatureMediaType
+	pluginConfig := opts.PluginConfig
+
+	trustpolicyDoc, err := v.TrustPolicyDocument()
+	if err != nil {
+		return notation.Descriptor{}, nil, err
+	}
+	trustPolicy, err := trustpolicyDoc.GetApplicableTrustPolicy(artifactRef)
+	if err != nil {
+		return notation.Descriptor{}, nil, notation.ErrorNoApplicableTrustPolicy{Msg: err.Error()}
+	}
+
+	outcome := &notation.VerificationOutcome{
+		VerificationResults: []*notation.ValidationResult{},
+		VerificationLevel:   opts.VerificationLevel,
+	}
+	err = v.processSignature(ctx, signature, envelopeMediaType, trustPolicy, pluginConfig, outcome)
+	if err != nil {
+		outcome.Error = err
+		return notation.Descriptor{}, outcome, err
+	}
+
+	payload := &envelope.Payload{}
+	err = json.Unmarshal(outcome.EnvelopeContent.Payload.Content, payload)
+	if err != nil {
+		outcome.Error = err
+		return notation.Descriptor{}, outcome, err
+	}
+	outcome.SignedAnnotations = payload.TargetArtifact.Annotations
+
+	return payload.TargetArtifact, outcome, nil
```

Comment:
Had a discussion earlier with @shizhMSFT on this one. For verifier.Verify(), the upstream caller should verify if the payload.TargetArtifact matches the actual target artifact being verified. notation.Verify() function contains this logic.
For users using verifier.Verify() only, this last step should be verified outside notation-go.

Reply:
I am worried that user of librart might miss to do that and consider verification success as overall success.

Reply:
> I am worried that user of library might miss to do that and consider verification success as overall success.

verifier.Verify() only verifies the signature by design. I added a comment in verifier.Verify()'s func doc to address this.
(Note: verifier.Verify() does not have repo related information. In other words, we are not able to resolve artifact reference to its descriptor inside this function.)

Reply:
I agree with @priteshbandi here. This is assuming best intentions from the library callers. Can we take the artifact descriptor as an input for this method?

Reply:
I agree with @priteshbandi and @rgnote. Since library users are not security experts, they may think that the artifact is verified after a successful call to the `Verifier.Verify()` even though they did not pass the artifact to the `Verify()` function.
> [Murphy's law](https://en.wikipedia.org/wiki/Murphy's_law): Anything that can go wrong will go wrong.

How about 
```golang
type Verifier interface {
    Verify(ctx context.Context, desc ocispec.Descriptor, signature []byte, opts notation.VerifyOptions) (*notation.VerificationOutcome, error)
}
```

Reply:
Updated the interface.